### PR TITLE
Optimizations: column encodings

### PR DESCRIPTION
**Pull request by martindurant**
Times for 10M values:

On Master:
category: read, no nulls 310.246
category: read, no nulls, has_null=True 316.878
category: read, with null, has_null=False 306.243
category: read, with null, has_null=True 310.223
category: write, no nulls 141.021
category: write, no nulls, has_null=True 535.763
category: write, with null, has_null=False 122.86
category: write, with null, has_null=True 544.889

Latest in this branch:
category: read, no nulls 50.333
category: read, no nulls, has_null=True 48.35
category: read, with null, has_null=False 49.018
category: read, with null, has_null=True 50.351
category: write, no nulls 13.652
category: write, no nulls, has_null=True 113.637
category: write, with null, has_null=False 14.475
category: write, with null, has_null=True 105.193

### CODE REVIEW

**Comment by mrocklin** [comment]
Awesome boost on the times.  What does this come from?

**Comment by martindurant** [comment]
The speed comes from not bitpacking the data with the minimum number of bits required to represent the biggest value, but rounding up to 8, so that numpy tostring() is the same as the packing. 
You may well be right about the case that the category count is > 2**8.

File: fastparquet/test/test_writer.py [hunk 1]
```diff
@@ -343,11 +343,6 @@ def test_empty_groupby(tempdir):
 
 
 
-@pytest.mark.parametrize('compression', ['GZIP',
-                                         'gzip',
-                                         None,
-                                         {'x': 'GZIP'},
-                                         {'y': 'gzip', 'x': None}])
```
Comment (martindurant) [inline fastparquet/test/test_writer.py]: How did that happen? :|

**Comment by martindurant** [comment]
Full set of benchmarks

Before
category: read, no nulls 304.37
category: read, no nulls, has_null=True 304.314
category: read, with null, has_null=False 303.338
category: read, with null, has_null=True 310.885
category: write, no nulls 125.793
category: write, no nulls, has_null=True 526.682
category: write, with null, has_null=False 121.521
category: write, with null, has_null=True 553.473
float64: read, no nulls 86.008
float64: read, no nulls, has_null=True 106.877
float64: read, with null, has_null=False 89.829
float64: read, with null, has_null=True 117.52
float64: write, no nulls 239.294
float64: write, no nulls, has_null=True 277.941
float64: write, with null, has_null=False 192.362
float64: write, with null, has_null=True 229.992
timedelta64[ns]: read, no nulls 113.296
timedelta64[ns]: read, no nulls, has_null=True 135.134
timedelta64[ns]: read, with null, has_null=False 116.96
timedelta64[ns]: read, with null, has_null=True 136.484
timedelta64[ns]: write, no nulls 216.5
timedelta64[ns]: write, no nulls, has_null=True 298.587
timedelta64[ns]: write, with null, has_null=False 237.286
timedelta64[ns]: write, with null, has_null=True 264.443
uint8: read, no nulls 22.033
uint8: read, no nulls, has_null=True 34.512
uint8: write, no nulls 223.083
uint8: write, no nulls, has_null=True 317.581

Latest
category: read, no nulls 4.94
category: read, no nulls, has_null=True 5.484
category: read, with null, has_null=False 4.471
category: read, with null, has_null=True 3.831
category: write, no nulls 14.723
category: write, no nulls, has_null=True 103.451
category: write, with null, has_null=False 16.625
category: write, with null, has_null=True 109.373
float64: read, no nulls 89.193
float64: read, no nulls, has_null=True 116.978
float64: read, with null, has_null=False 91.357
float64: read, with null, has_null=True 108.176
float64: write, no nulls 87.879
float64: write, no nulls, has_null=True 191.524
float64: write, with null, has_null=False 132.795
float64: write, with null, has_null=True 170.687
timedelta64[ns]: read, no nulls 119.173
timedelta64[ns]: read, no nulls, has_null=True 147.922
timedelta64[ns]: read, with null, has_null=False 117.955
timedelta64[ns]: read, with null, has_null=True 146.988
timedelta64[ns]: write, no nulls 211.171
timedelta64[ns]: write, no nulls, has_null=True 218.825
timedelta64[ns]: write, with null, has_null=False 168.874
timedelta64[ns]: write, with null, has_null=True 224.411
uint8: read, no nulls 26.04
uint8: read, no nulls, has_null=True 42.306
uint8: write, no nulls 60.194
uint8: write, no nulls, has_null=True 167.039

The results look too good for categories, this is really bare metal.

**Comment by martindurant** [comment]
Latest:
category: read, no nulls 4.937
category: read, no nulls, has_null=True 4.764
category: read, with null, has_null=False 3.329
category: read, with null, has_null=True 3.315
category: write, no nulls 22.944
category: write, no nulls, has_null=True 105.005
category: write, with null, has_null=False 15.051
category: write, with null, has_null=True 102.041
float64: read, no nulls 34.652
float64: read, no nulls, has_null=True 60.941
float64: read, with null, has_null=False 36.534
float64: read, with null, has_null=True 60.727
float64: write, no nulls 97.683
float64: write, no nulls, has_null=True 143.593
float64: write, with null, has_null=False 147.014
float64: write, with null, has_null=True 171.489
timedelta64[ns]: read, no nulls 65.265
timedelta64[ns]: read, no nulls, has_null=True 92.912
timedelta64[ns]: read, with null, has_null=False 63.268
timedelta64[ns]: read, with null, has_null=True 89.366
timedelta64[ns]: write, no nulls 163.062
timedelta64[ns]: write, no nulls, has_null=True 208.022
timedelta64[ns]: write, with null, has_null=False 163.421
timedelta64[ns]: write, with null, has_null=True 214.995
uint8: read, no nulls 8.306
uint8: read, no nulls, has_null=True 8.833
uint8: write, no nulls 23.981
uint8: write, no nulls, has_null=True 107.991

These are all for single row-group, single column, so some acceleration will not pass through to more typical use. I still don't know how the category column is quite so fast, but appears to work.

**Pull request merged**